### DISCUSSION TITLE
explicit "extension + using static" on methods

### DISCUSSION THREAD

**Discussion by ZacharyPatten** [dotnet/csharplang#5105] [Language Ideas]
It would be nice if there was a way to allow methods to both be an extension and support `using static ...;` syntax. I'm under the impression it was a design descision not to allow extension methods to be called from `using static ...;` syntax, but there should be a way to explicitly opt-into being able to do so. Currently the following does not compile:
```cs
using static Stuff1;

0.Method();
Method(0); // <- CS0103

public static class Stuff1
{
    /// <summary>hello world</summary>
    public static void Method(this int a) { /* some code */ }
}
```

Because there is currently no way to allow a method to be both an extension and support `using static ...;` it means if you want to support it you have to make a new method in a seperate type:
```cs
using static Stuff2;

0.Method(); // <- "Stuff1"
Method(0);  // <- "Stuff2"

public static class Stuff1
{
    /// <summary>hello world</summary>
    public static void Method(this int a) { /* some code */ }

    // public static void Method(int a) { } // <- CS0121 if uncommented
}

public static class Stuff2
{
    /// <inheritdoc cref="Stuff1.Method(int)"/>
    public static void Method(int a) => a.Method();
}
```

Can we just add a way to explicitly do this without having to create multiple types and methods? Here are a few possible ideas...

- via attribute:
```cs
using static Stuff1;

0.Method();
Method(0);

public static class Stuff1
{
    /// <summary>hello world</summary>
    [AllowUsingStatic] public static void Method(this int a) { /* some code */ }
}
```

- via new syntax (`this?` to indicate calling as an extension is optional):
```cs
using static Stuff1;

0.Method();
Method(0);

public static class Stuff1
{
    /// <summary>hello world</summary>
    public static void Method(this? int a) { /* some code */ }
}
```

- via using directive syntax (Instead of syntax on each method, explicit syntax on the using directive)
```cs
using all static Stuff1; // I dont like "all" but some verbage here

0.Method();
Method(0);

public static class Stuff1
{
    /// <summary>hello world</summary>
    public static void Method(this int a) { /* some code */ }
}
```

Additional Note:
The example I shared was very simple. As methods have more parameters and generic parameters, requiring duplicate methods gets more verbose.

**Comment by bernd5** [comment]
I found this: https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2014/LDM-2014-02-10.md#design-of-using-static

**Reply by ZacharyPatten** [reply]
Thanks for the link.

I disagree with the reasoning specified in that link:

> you only call them as static methods to disambiguate

I personally prefer to call them as static. I often only make extension methods to make it easier for other developers to see them (autocomplete on instances).

The decision was already made though, so like it or not we are stuck with it... but  we can add a way to explicitly prefer the opposite of that opinion (as this discussion is asking for).

**Comment by Rekkonnect** [comment]
Extension methods are literally syntax sugar, in reality they're static methods, that must be declared in static classes for them to be used as extension methods. Nothing prevents you from using them as normal static methods, unless you're importing them through `using static`.

The decision for this exception, which is shown in the LDM notes that are linked in another comment, doesn't seem to handle an erroneous case. Instead, this restriction may be freely lifted and permit extension methods being imported through a `using static` statement. This would however break existing code in certain occasions, which will be resolved by first attempting to bind a method invocation to a normal static method, and then an extension one, if binding fails.

I may be naively ignoring other repercussions, which I'd gladly consider.
